### PR TITLE
DataFormats/CaloTowers : changes needed to compile on macOS sierra with system clang and libc++

### PR DESCRIPTION
**Pull request by gartung**
input_line_480:5:39: error: rvalue reference to type 'vector<...>' cannot bind to lvalue of type 'vector<...>'
    ((CaloTower*)obj)->setConstituents((vector<DetId>&)*(vector<DetId>*)args[0]);
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-az5p3l77zcgnhp5tqilkw3riti3zbviu/CMSSW_9_0_X/src/DataFormats/CaloTowers/interface/CaloTower.h:65:46: note: passing argument to parameter 'ids' here
   void setConstituents( std::vector<DetId>&& ids ) { constituents_=std::move(ids);}
                                              ^
Error in <TClingCallFunc::make_wrapper>: Failed to compile

https://sft.its.cern.ch/jira/browse/ROOT-8075

### CODE REVIEW

File: DataFormats/CaloTowers/interface/CaloTower.h [hunk 1]
```diff
@@ -62,7 +62,7 @@ class CaloTower : public reco::LeafCandidate {
    // setters
   void addConstituent( DetId id ) { constituents_.push_back( id ); }
   void addConstituents( const std::vector<DetId>& ids );
-  void setConstituents( std::vector<DetId>&& ids ) { constituents_=std::move(ids);}
+  void setConstituents( std::vector<DetId>& ids ) { constituents_=ids;}
```
Comment (slava77) [inline DataFormats/CaloTowers/interface/CaloTower.h]: If I understand correctly, there is extra cost to this,
to be evaluated.
If needed in cling, appropriate ifdefs can deliver needed content.
Also, there should be a const qualifier here

**Comment by gartung** [comment]
Correct. This is for complaints from rootcling when selecting certain collections in cmsShow.exe. I will wrap this with #ifdef __ROOTCLING__.

File: DataFormats/CaloTowers/interface/CaloTower.h [hunk 2]
```diff
@@ -62,8 +62,11 @@ class CaloTower : public reco::LeafCandidate {
    // setters
   void addConstituent( DetId id ) { constituents_.push_back( id ); }
   void addConstituents( const std::vector<DetId>& ids );
+#ifdef __ROOTCLING__
+  void setConstituents( std::vector<DetId>& ids ) { constituents_=ids;}
```
Comment (slava77) [inline DataFormats/CaloTowers/interface/CaloTower.h]: I still think that const as in ```setConstituents( const std::vector<DetId>& ids ) ``` is needed here

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17380	c285eb3
- technical change as described (hide rvalue ref interface from cling)
- jenkins tests pass

**Comment by davidlange6** [comment]
+1

**Pull request merged**